### PR TITLE
DOCSP-4304: Validate rst structure

### PR DESCRIPTION
**Pull request by i80and**
This replaces the ad-hoc unstructured handling of directives and roles with a more rigorous TOML definition. Not everything is currently validated; for example, `content_type` is technically treated as a boolean. But the intent is for us to be able to ensure at parse-time that anything with a `content_type="list-table"` has a `list-table`-style structure.

Any thoughts appreciated!

### CODE REVIEW

File: snooty/rstparser.py [hunk 1]
```diff
@@ -365,13 +287,64 @@ def add_diagnostics(self, diagnostics: Iterable[Diagnostic]) -> None: ...
...
+            option_name:
+                spec.get_validator(option) for option_name, option in directive.options.items()
+        }
+
+        class DocutilsDirective(BaseDocutilsDirective):
+            has_content = bool(directive.content_type)
+            optional_arguments = 1 if directive.argument_type else 0
+            final_argument_whitespace = True
+            option_spec = options
+            snooty_directive = directive
```
Comment (dpercy) [inline snooty/rstparser.py]: This field looks unused

File: snooty/specparser.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+        elif isinstance(option_spec, str) and option_spec in self.enum:
+            return lambda argument: docutils.parsers.rst.directives.choice(
+                argument,
+                self.enum[cast(str, option_spec)])
+
+        raise ValueError(f'Unknown directive argument type "{option_spec}"')
+
+    def _resolve_inheritance(self) -> None:
+        """Directives can inherit from other directives; resolve this."""
+        def resolve(inheritable_index: Dict[str, _T]) -> None:
```
[the remark this answers is not in the document]
Comment (i80and) [reply snooty/specparser.py]: :+1:

File: snooty/test_specparser.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
+from . import specparser
+from typing import Any, Callable, Type, Optional
+
+
+def ensure_failure(func: Callable[[], Any], exception_class: Type[Exception]) -> None:
```
[the remark this answers is not in the document]
Comment (i80and) [reply snooty/test_specparser.py]: :+1:

File: snooty/rstparser.py [hunk 4]
```diff
@@ -256,7 +196,7 @@ def run(self) -> List[docutils.nodes.Node]:
             return [node]
 
         # The new syntax needs no special handling
-        return Directive.run(self)
+        return BaseDocutilsDirective.run(self)
```
[the remark this answers is not in the document]
Comment (i80and) [reply snooty/rstparser.py]: :+1:

File: snooty/rstparser.py (hunk 1 above)
Comment (i80and) [reply snooty/rstparser.py]: :+1:

**Comment by i80and** [comment]
The reason directive/role/rstobject definitions are in a language-agnostic format is to decouple as much as possible from Python so that we can reuse this data in the frontend or even, in the future, from a completely different parser written in, for example, Rust

**Comment by i80and** [comment]
Additionally, we want to be able to decouple spec changes from new parser releases and have snooty-parser download the latest spec once a day or so. Download and running arbitrary Python has much more stringent of security demands than downloading and parsing a TOML file

**Pull request merged**